### PR TITLE
Add `load_state_dict` error message for non-dicts

### LINKED ISSUE #76886: More informative load_state_dict error message
### 🚀 The feature, motivation and pitch

Currently, feeding a string into `torch.nn.Module.load_state_dict` results in an undescriptive exception (`AttributeError` in 1.10.2, `ValueError` in the current version). Checking for the object's type before passing the dict to hooks and raising a TypeError with an appropriate message would make the source of the error clearer, especially for new users.

### Alternatives

Adding an assert; using `typeguard`'s decorator

### Additional context

_No response_

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345

### PR DESCRIPTION
**Pull request by neverix**
Fixes #76886

### CODE REVIEW

**Review by jbschlosser — COMMENTED** [review]
Thanks for the contribution! Do you mind adding a test that demonstrates the error message throwing behavior when a non-dict-like is passed to `load_state_dict()`? A good place to put it is just after this test in `test/test_nn.py`:
https://github.com/pytorch/pytorch/blob/61dcde88a6c2ce60000b59e8e472a9a6c2456f2c/test/test_nn.py#L1403-L1414

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -1,4 +1,5 @@
 from collections import OrderedDict, namedtuple
+import collections.abc
```
[the remark this answers is not in the document]
Comment (neverix) [reply torch/nn/modules/module.py]: Would it help to add tests for the error not throwing when a normal dict is passed? Not sure how to match a regular try/except exception with regex.

Though that part is already checked by other tests.

[the remark this answers is not in the document]
Comment (jbschlosser) [reply torch/nn/modules/module.py]: Yeah exactly, I'd expect that to be covered by other tests. Ideally, none of those should fail as a result of this addition.

**Review by jbschlosser — APPROVED** [review]
LGTM thanks!

**Pull request closed without merging**